lectures/zksnark.tex

```diff
@@ -0,0 +1,812 @@
+    \caption{AND Gate Truth Table}
+    \label{tab:and-truth-table}
+    \vspace{1em}
+\end{wraptable}
+
+% --- Finish Writing diagrams ---
+
+The \textit{AND Gate Truth \Cref{tab:and-truth-table}} shows us the results we receive if 
+particular values are supplied to the gate. The main point here is that with this table, we can 
+verify the validity of logical statements. Boolean circuits receive an input vector of $\{0, 1\}$ 
+and resolve to true (1) or false (0); basically, they determine if the input values satisfy the 
+statement.
+
+We can do the same with \textbf{arithmetic circuits} to verify computations without excessive 
+verbosity because of binary arithmetic.
+
+Let's take a look at some examples of programs and how can we translate them to the arithmetic
+circuits.
+
+Very simple program with a multiplication.
+
+\begin{lstlisting}[language=C,numbers=none]
+    return a * b
+\end{lstlisting}
+
+This can be represented as a circuit with only one gate:
+\[ r = a \times b \]
+
+The witness vector (solution vector) is $w = (r, a, b)$, for example: $(6, 2, 3)$. We 
+assume that the $a$ and $b$ are input values. 
+
+We can think of the ``=`` in the gate as an assertion, meaning that if $a \times b$ does not equal
+$r$, the assertion fails, and the input values don't resolve the circuit.
+
+How can we translate an if statement?
+
+\begin{lstlisting}[language=C,numbers=none]
+    if (a) {
+        return b * c
+    } else {
+        return b + c
+    }
+\end{lstlisting}
+
+We can express this logic in mathematical terms as follows: "If $a$ is true, compute 
+$b \times c$; otherwise, compute $b + c$." Only numerical expressions are allowed. Knowing that
+$\text{true} := 1$ and $\text{false} := 0$, we can transform it as follows:
+\[ r = a \times (b \times c) + (1 - a) \times (b + c) \]
+Where the witness vector is: $w = (r, a , b, c)$: $(6, 1, 2, 3)$, $(5, 0, 2, 3)$.\\
+
+But, we need to verify all the intermediate steps. This can be achieved by transforming the above
+equation using the simplest terms (the gates), ensuring the correctness of each step in the program: 
+
+\begin{equation*}
+    \begin{aligned}
+        r_1 &= b \times c \\
+        r_2 &= b + c \\
+        r_3 &= 1 - a \\
+        r_4 &= a \times r_1 \\
+        r_5 &= r_3 \times r_2 \\
+        r &= r_4 + r_5 \\
+    \end{aligned}
+\end{equation*}
+
+With witness vector: $w = (r, r_1, r_2, r_3, r_4, r_5, a , b, c)$: $(6, 6, 5, 0, 6, 0, 1, 2, 3)$.
+
+\subsection{Rank-1 Constraint System}
+
+Almost any program written in high-level programming language can be translated (compiled) into
+arithmetic circuits, that are really powerfull tool. But for the ZK proof we need slightly different
+format of it - \textbf{Rank-1 Constraint System}, where the simpliest term is \textbf{constraint}. 
+This offers a more flexible and general way to describe these parts.
+    
+\begin{definition}
+    Let $\mathbb{V}$ be a vector space over the field $\mathbb{F}$. An \textbf{inner product} 
```

Comment:
Well, I would rephrase this definition a bit. 

The inner product of a linear space $\mathbb{V}$ is **any** symmetric, linear in the first argument, and positive binary function from vector space to a set of scalars. The specific function we are used to $\langle u,v \rangle = \sum_{i=1}^n u_iv_i$ (which is formally not a linear combination) is just a special case of this general definition. You wrote everything right, but formally the properties should go before the specific function we are using in SNARKs.

Reply:
One example of another valid scalar product formulation for $\mathbb{R}^n$ is $\langle u,v \rangle := \sum_{i=1}^n \mu_iu_iv_i$ for some positive set $\\{\mu_i\\}_{i=1}^n \subset \mathbb{R}\_{>0}$.

Moreover, any symmetric positive-definite matrix $M$ would make $\langle u,v \rangle = u^{\top}Mv$ a valid scalar product! :) But that is just for you to know, no need to insert it to make matters more complicated

Reply:
Will it be okay as written below?

The inner product of a linear space $\mathbb{V}$ is any symmetric, linear in the first 
argument, and positive binary function from the vector space to a set of scalars.

Let $\mathbb{V}$ be a vector space over the field $\mathbb{F}$. An **inner product** 
on $\mathbb{V}$ is a function:

$$
\langle \cdot, \cdot \rangle: \mathbb{V} \times \mathbb{V} \rightarrow \mathbb{F}
$$

defined for $u, v \in \mathbb{V}$ as follows:

$$
\langle u, v \rangle = \sum_{i=1}^{n} u_i v_i
$$

The basic properties that must be satisfied for all $c \in \mathbb{F}$ and for all $u, v, w \in \mathbb{V}$ are:

- **Symmetry:** $\langle u, v \rangle = \langle v, u \rangle$
- **Linearity in the first argument:** $\langle cu + v, w \rangle = c \langle u, w \rangle + \langle v, w \rangle$
- **Positivity:** $\langle u, u \rangle \geq 0$ and $\langle u, u \rangle = 0 \iff u = 0$